### ISSUE TITLE
Adversarial backend projection

### ISSUE THREAD

**Issue by alexzfan** [fairlearn/fairlearn#1234]
I noticed that the projection code is different b/w the pytorch and the tensorflow backend. 

tf: `proj = tensorflow.reduce_sum(tensorflow.multiply(dW_LP[i], unit_dW_LA))` does an elementwise mult + sum.
torch: `proj = torch.sum(torch.inner(unit_dW_LA, dW_LP[i]))` does some kind of tensordot operation for matrices (at least according to torch docs) and then sums. 

Seems like tf is implemented correctly, but would like confirmation

**Comment by hildeweerts** [comment]
Hi @alexzfan, thanks for opening the issue! 

I have to admit my linear algebra is a bit rusty, but I believe that for **matrices** the sum of an element-wise multiplication is **not** equal to the dot product, but for two **column vectors** the result should be the equivalent (though in that case `torch.sum` doesn't really add much because the dot product already takes care of summation).

@SeanMcCarren could you perhaps chime in here?

**Comment by alexzfan** [comment]
Hi @hildeweerts and @SeanMcCarren , for some additional context here, the main issue arises when training a large convnet like resnet in torch. 

If we have some arbitrary convolution grad weights matrices:
`a = torch.randn(128, 128, 3,3)
b = torch.randn(128,128,3,3)
torch.inner(a,b).shape`

then the resulting tensor has shape (128,128,3,128,128,3), which overloads GPU memory before the contents are able to be summed.

**Comment by SeanMcCarren** [comment]
@alexzfan those two lines are indeed different, for the reason hilde mentioned. Inner product and multiplication are in general very different.

About your latest post, I'm not sure what your are trying to do? I believe you're not sure what inner product does? personally I don't even know how to define inner product on arbitrary dimensional tensors but torch does. I recommend just go ogling for videos about it. :) Anyways, I think you want to do convolutions, so here are the operators for that  https://pytorch.org/docs/stable/nn.functional.html#convolution-functions

**Comment by alexzfan** [comment]
@SeanMcCarren thanks for the response! Let me just clarify my thoughts then.

1. If the element-wise multiplication is incorrect for the tf implementation, then I think that’s a bug that’s present in the current tensor flow backend engine repo code. I’m open to submitting another issue report about that just to keep things organized.

2. My other question is more about actually using the adversarial debasing techniques for large convolutional networks that have large multi-dimensional gradients on the weights. I have trouble training them on GPU since the torch.inner function creates a (for lack of better term) enormous tensor that can’t be stored, so I was wondering if there’s some equivalent code that doesn’t store the intermediate output of torch.inner.

**Comment by SeanMcCarren** [comment]
Ah that is really clearly put, thanks! 

In this case the implementation is indeed not suitable for your purpose. The default implementation of a model in the tf and torch backends were made with only simple layers in mind, not convolutions. You can imagine that you can make the model as complex as you'd like, but it isn't sensible for us to support all of it. So about question 1, I don't think you need to, but good initiative :) 

Instead, we support you creating your own model! Hence, I think the best approach for you is to create your own tf model containing these convolutions and pass it directly when initializing AdversarialFairness.

If you are unsure how to do this, please drop a message with your model and maybe I can help!

**Comment by alexzfan** [comment]
@SeanMcCarren gotcha, thanks for the help! I actually emailed the original author of the paper, and he explained that the weights should just be taken as giant vectors when doing the projection (which I think means to flatten the tensor, project, and reshape the tensor back into the original grad weight shape). I'll just adapt the backend code for my project with that idea.

**Comment by SeanMcCarren** [comment]
Do you mean for the convolutional layers? Because only flattening is not how convolution works I'm afraid!

I still recommend creating the tf model yourself, it's quite fun. Maybe [https://www.tensorflow.org/tutorials/images/CNN](this tutorial) helps
